scheduler/cloudformation/cloudformation.go

```diff
@@ -1,5 +1,554 @@
+	Cluster string
+
+	// If true, wait for stack updates and creates to complete.
+	Wait bool
+
+	// The name of the bucket to store templates in.
+	Bucket string
+
+	// CloudFormation client for creating stacks.
+	cloudformation cloudformationClient
+
+	// ECS client for performing ECS API calls.
+	ecs ecsClient
+
+	// S3 client to upload templates to s3.
+	s3 s3Client
+
+	db *sql.DB
+}
+
+// NewScheduler returns a new Scheduler instance.
+func NewScheduler(db *sql.DB, config client.ConfigProvider) *Scheduler {
+	return &Scheduler{
+		cloudformation: cloudformation.New(config),
+		ecs:            ecs.New(config),
+		s3:             s3.New(config),
+		db:             db,
+	}
+}
+
+// Submit creates (or updates) the CloudFormation stack for the app.
+func (s *Scheduler) Submit(ctx context.Context, app *scheduler.App) error {
+	tx, err := s.db.Begin()
+	if err != nil {
+		return err
+	}
+
+	err = s.submit(ctx, tx, app)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	return tx.Commit()
+}
+
+// Submit creates (or updates) the CloudFormation stack for the app.
+func (s *Scheduler) submit(ctx context.Context, tx *sql.Tx, app *scheduler.App) error {
+	stackName, err := s.stackName(app.ID)
+	if err == errNoStack {
+		stackName = app.Name
+		if _, err := tx.Exec(`INSERT INTO stacks (app_id, stack_name) VALUES ($1, $2)`, app.ID, stackName); err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	}
+
+	buf := new(bytes.Buffer)
+	if err := s.Template.Execute(buf, app); err != nil {
+		return err
+	}
+
+	key := fmt.Sprintf("%x", sha1.Sum(buf.Bytes()))
+
+	_, err = s.s3.PutObject(&s3.PutObjectInput{
+		Bucket:      aws.String(s.Bucket),
+		Key:         aws.String(fmt.Sprintf("/%s", key)),
+		Body:        bytes.NewReader(buf.Bytes()),
+		ContentType: aws.String("application/json"),
+	})
+	if err != nil {
+		return err
+	}
+
+	url := fmt.Sprintf("https://%s.s3.amazonaws.com/%s", s.Bucket, key)
+
+	desc, err := s.cloudformation.DescribeStacks(&cloudformation.DescribeStacksInput{
+		StackName: aws.String(stackName),
+	})
+
+	tags := []*cloudformation.Tag{
+		{Key: aws.String("empire.app.id"), Value: aws.String(app.ID)},
+		{Key: aws.String("empire.app.name"), Value: aws.String(app.Name)},
+	}
+
+	if err, ok := err.(awserr.Error); ok && err.Message() == fmt.Sprintf("Stack with id %s does not exist", stackName) {
+		if _, err := s.cloudformation.CreateStack(&cloudformation.CreateStackInput{
+			StackName:   aws.String(stackName),
+			TemplateURL: aws.String(url),
+			Tags:        tags,
+		}); err != nil {
+			return err
+		}
+
+		if s.Wait {
+			if err := s.cloudformation.WaitUntilStackCreateComplete(&cloudformation.DescribeStacksInput{
+				StackName: aws.String(stackName),
+			}); err != nil {
+				return err
+			}
+		}
+	} else if err == nil {
+		stack := desc.Stacks[0]
+		status := *stack.StackStatus
+
+		// If there's currently an update happening, wait for it to
+		// complete.
+		if strings.Contains(status, "IN_PROGRESS") {
```

Comment:
Yeah, I waffled on that. I figure that that cause should be pretty rare, and it can be handled manually by resolving the issue that caused the rollback, deleting the stack, then re-deploying.